File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
DNA = 'ACGT'
num = [0, 0, 0, 0]

S, P = map(int, input().split())
sequence = input()
min_num = list(map(int, input().split()))
answer = 0

for i in range(P):
    word = sequence[i]
    num[DNA.index(word)] += 1

is_true = True

for i in range(4):
    if min_num[i] > num[i]:
        is_true = False
        break

if is_true:
    answer += 1

for i in range(S - P):
    word = sequence[i]
    num[DNA.index(word)] -= 1
    word = sequence[P + i]
    num[DNA.index(word)] += 1

    is_true = True

    for i in range(4):
        if min_num[i] > num[i]:
            is_true = False
            break

    if is_true:
        answer += 1

print(answer)

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.cc
#include <iostream>
using namespace std;

char DNA[1000001];
int min_num[4];
int counts[4];

int check_sequence() {
    int result = 1;

    for (int i = 0; i < 4; i++)
    {
        if (min_num[i] > counts[i]) {
            result = 0;
            break;
        }
    }

    return result;
}

int check_base(char base) {
    if (base == 'A') return 0;
    else if (base == 'C') return 1;
    else if (base == 'G') return 2;
    else return 3;
}

int main()
{
    int S_len, P_len, temp;
    cin >> S_len >> P_len;
    cin >> DNA;

    // A C G T
    for (int i = 0; i < 4; i++)
    {
        cin >> min_num[i];
    }

    char ACGT[4] = { 'A', 'C', 'G', 'T' };

    for (int i = 0; i < P_len; i++)
    {
        temp = check_base(DNA[i]);
        counts[temp] += 1;
    }

    int answer = 0;

    if (check_sequence() == 1) {
        answer += 1;
    }
     
    for (int i = 0; i < S_len - P_len; i++)
    {
        temp = check_base(DNA[i]);
        counts[temp] -= 1;
        temp = check_base(DNA[i + P_len]);
        counts[temp] += 1;

        if (check_sequence() == 1) {
            answer += 1;
        }
    }

    cout << answer;

    return 0;
}
